torch/_dynamo/symbolic_convert.py

```diff
@@ -2628,6 +2628,55 @@ def __init__(
         )
         linecache.lazycache(f_code.co_filename, f_globals)
 
+    def _init_torch_function_state(self):
+        # This is annoyingly complicated because of how the torch function subclass + mode C API was designed
+        # There are two exposed C knobs here as contexts: torch._C.DisableTorchFunction and torch._C.DisableTorchFunctionSubclass
+        # These are their definitions:
+        # 1) torch._C._is_torch_function_enabled indicates that neither of the above knobs have been entered
+        # (if either are entered, this will be False)
+        # 2) torch._C._is_torch_function_mode_enabled indicates that either the torch mode stack is empty OR
+        # torch._C.DisableTorchFunction has been entered
+        # To disambiguate these and keep myself sane I added a C API to check whether all torch function
+        # concepts (modes and subclasses) are enabled.
+        # This only returns true iff we have not entered torch._C.DisableTorchFunction and allows us to separate
+        # the stack length from the enablement state of torch function modes.
+        # This is important because now if a mode is pushed while dynamo is tracing, we know whether
+        # or not torch function modes are enabled and whether we should trace it.
+        self.torch_function_subclass_enabled = torch._C._is_torch_function_enabled()
+
+        # This differs from the C API of the same name
+        # this will only be false iff we have entered torch._C.DisableTorchFunction
+        # and does not take into account the mode stack length, while the C API bundles these
+        # two concepts
+        self.torch_function_mode_enabled = (
+            not torch._C._is_torch_function_all_disabled()
+        )
+
+        self.cur_mode = None
+
+    @contextlib.contextmanager
+    def _pop_mode_for_inlining(self):
```

Comment:
What does "_pop_mode_for_inlining" mean?

Reply:
Before running a torch function mode, it is popped off of the stack. This way, recursively calling `func` in the body of the mode's `__torch_function__` impl doesn't infinite loop. Each mode is handled in turn, and placed back on the stack in reverse order as the call stack of `__torch_function__` unwinds once a result is returned.